File: src/main/java/net/cofcool/toolbox/internal/simplenote/NoteService.java
package net.cofcool.toolbox.internal.simplenote;

import io.vertx.core.Future;
import io.vertx.core.Vertx;
import java.util.List;
import lombok.CustomLog;
import net.cofcool.toolbox.internal.simplenote.entity.Note;
import net.cofcool.toolbox.util.SqlRepository;

@CustomLog
public class NoteService {

    private final SqlRepository<Note> noteRepository;

    public NoteService(Vertx vertx) {
        noteRepository = SqlRepository.create(vertx, Note.class);
    }

    public Future<Note> save(Note note)  {
        var n = note;
        if (n.id() == null) {
            n = Note.init(n.content());
        }
        return noteRepository.save(n);
    }

    public Future<Void> save(List<Note> notes) {
        return noteRepository.save(notes);
    }

    Future<Void> delete(Note note) {
        return noteRepository.delete(note.id());
    }

    Future<Void> logicDelete(Note note) {
        return noteRepository.find(note.id())
            .compose(n ->
                noteRepository.save(new Note(n.id(), n.content(), n.date(), Note.NoteState.DELETE_FLAG.name()))
            )
            .compose(n -> n == null ? Future.failedFuture("Update error") : Future.succeededFuture());
    }

    Future<List<Note>> find(Note note) {
        if (note == null) {
            return noteRepository.find();
        } else {
            return noteRepository.find(note);
        }
    }

}

File: simple-note/src/main/kotlin/net/cofcool/toolbox/simplenote/internal/NoteService.kt
package net.cofcool.toolbox.simplenote.internal

import io.vertx.core.Future
import io.vertx.core.Vertx
import net.cofcool.toolbox.simplenote.Config
import net.cofcool.toolbox.simplenote.Note
import net.cofcool.toolbox.simplenote.NoteRepository
import net.cofcool.toolbox.simplenote.NoteState
import kotlin.io.path.Path


class NoteService(vertx: Vertx) {

    private val noteRepository: NoteRepository;

    init {
        val config = vertx.orCreateContext.config()
        val path = Path(
            config.getString(Config.PATH_KEY, Config.PATH_VAL),
            config.getString(Config.FILE_KEY, Config.FILE_NAME)
        )
            .toAbsolutePath().toString()
        noteRepository = FileRepository(path, vertx)
    }

    fun save(note: Note): Future<Note> {
        return Future.future() {
            it.complete(noteRepository.save(note))
        }
    }

    fun save(notes: List<Note>): Future<Void> {
        return Future.future() {
            noteRepository.save(notes)
            it.complete()
        }
    }

    fun delete(note: Note): Future<Void> {
        return Future.future {
            noteRepository.delete(note.id)
            it.complete()
        }
    }

    fun logicDelete(note: Note): Future<Void> {
        return Future.future {
            noteRepository.find(note.id)?.let {
                it.state = NoteState.DELETE_FLAG
                noteRepository.save(it)
            }
            it.complete()
        }
    }

    fun find(note: Note?): Future<List<Note>> {
        return Future.future {
            it.complete(
                if (note == null) {
                    noteRepository.find()
                } else {
                    noteRepository.find(note)
                }
            )
        }
    }

}
